### PR TITLE
[Terrain] Prevent AP from processing image until after it's done saving

### PR DESCRIPTION
**Pull request by mbalfour-amzn**
## What does this PR do?

When painting or baking an image gradient, it was possible for the Asset Processor to start processing the asset before the image was fully written, which would result in an incorrect processed asset. Furthermore, the AP didn't know to reprocess the asset without a manual trigger.

This change uses special temp file naming to make the AP ignore the image while it's being written, then renames it _after_ it's done so that the AP processes the fully-written image.

Also fixed a tiny null dereference problem in the image builder that could happen occasionally if the source image couldn't be read in for some reason.

## How was this PR tested?

Tried painting on images that were 64x64 in size and saving them. Oddly enough, tiny images like these were triggering the problem 100% of the time for me. They now trigger it 0% of the time.

### CODE REVIEW

**Review by lawsonamzn — APPROVED** [review]

File: Gems/GradientSignal/Code/Source/Editor/EditorGradientImageCreatorUtils.cpp [hunk 1]
```diff
@@ -211,11 +215,16 @@ namespace GradientSignal::ImageCreatorUtils
...
-            AZ_Error("EditorImageGradientComponent", result, "Failed to write out gradient image to path: %s", absolutePath.c_str());
+            AZ_Error("EditorImageGradientComponent", result, "Failed to write out gradient image to path: %s", tempSavePath.c_str());
         }
 
         outputImage->close();
 
+        // Now that we're done saving the temporary image, rename it to the correct file name.
+        [[maybe_unused]] auto moveResult = AZ::IO::SmartMove(tempSavePath.c_str(), absolutePath.c_str());
+        AZ_Error("EditorImageGradientComponent", moveResult,
+            "Failed to rename temporary image asset %s", tempSavePath.c_str(), absolutePath.c_str());
```
Comment (invertednormal) [inline Gems/GradientSignal/Code/Source/Editor/EditorGradientImageCreatorUtils.cpp:226]: You're passing both path strings into this message, but it only takes one.

**Review by invertednormal — CHANGES_REQUESTED** [review]
Just one small thing needs to be fixed.

File: Gems/GradientSignal/Code/Source/Editor/EditorGradientImageCreatorUtils.cpp (hunk 1 above)
Comment (mbalfour-amzn) [reply Gems/GradientSignal/Code/Source/Editor/EditorGradientImageCreatorUtils.cpp:226]: Whoops ty! It surprisingly didn't complain!

Comment (lawsonamzn) [reply Gems/GradientSignal/Code/Source/Editor/EditorGradientImageCreatorUtils.cpp:226]: nicely spotted!

**Review by cgalvan — APPROVED** [review]

**Comment by cgalvan** [comment]
@mbalfour-amzn  Does this fix https://github.com/o3de/o3de/issues/13064 ?

**Review by invertednormal — APPROVED** [review]

File: Gems/GradientSignal/Code/Source/Editor/EditorGradientImageCreatorUtils.cpp [hunk 2]
```diff
@@ -211,11 +215,16 @@ namespace GradientSignal::ImageCreatorUtils
...
         {
-            AZ_Error("EditorImageGradientComponent", result, "Failed to write out gradient image to path: %s", absolutePath.c_str());
+            AZ_Error("EditorImageGradientComponent", result, "Failed to write out gradient image to path: %s", tempSavePath.c_str());
         }
 
         outputImage->close();
 
+        // Now that we're done saving the temporary image, rename it to the correct file name.
+        [[maybe_unused]] auto moveResult = AZ::IO::SmartMove(tempSavePath.c_str(), absolutePath.c_str());
+        AZ_Error("EditorImageGradientComponent", moveResult,
```
Comment (jackalbe) [inline Gems/GradientSignal/Code/Source/Editor/EditorGradientImageCreatorUtils.cpp:225]: Should "moveResult" be used in the return value?

Comment (mbalfour-amzn) [reply Gems/GradientSignal/Code/Source/Editor/EditorGradientImageCreatorUtils.cpp:225]: Yep, good catch!

**Pull request merged**